spec/support/fake_analytics_spec.rb

```diff
@@ -0,0 +1,350 @@
+        expect { expect(subject).to(expectation.call) }.
+          to raise_error(RSpec::Expectations::ExpectationNotMetError) do |err|
+          assert_error_messages_equal(err, <<~MESSAGE)
+            Expected that FakeAnalytics would have received event nil
+            with nil.
+  
+            Events received:
+            {}
+          MESSAGE
+        end
+      end
+
+      it 'does not raise if event was logged 1x' do
+        track_event.call
+        expect { expect(subject).to(expectation.call) }.
+          not_to raise_error(RSpec::Expectations::ExpectationNotMetError)
+      end
+
+      it 'does not raise if event was logged 2x' do
+        track_event.call
+        track_event.call
+
+        expect { expect(subject).to(expectation.call) }.
+          not_to raise_error(RSpec::Expectations::ExpectationNotMetError)
+      end
+    end
+
+    context 'event name only' do
+      let(:track_event) { -> { subject.track_event :my_event } }
+      let(:expectation) { -> { have_logged_event :my_event } }
+
+      it 'raises if no event has been logged' do
+        expect { expect(subject).to(expectation.call) }.
+          to raise_error(RSpec::Expectations::ExpectationNotMetError) do |err|
+            assert_error_messages_equal(err, <<~MESSAGE)
+              Expected that FakeAnalytics would have received event :my_event
+              with nil.
+
+              Events received:
+              {}
+            MESSAGE
+          end
+      end
+
+      it 'raises if another type of event has been logged' do
+        subject.track_event(:my_other_event)
+
+        expect { expect(subject).to(expectation.call) }.
+          to raise_error(RSpec::Expectations::ExpectationNotMetError) do |err|
+            assert_error_messages_equal(err, <<~MESSAGE)
+              Expected that FakeAnalytics would have received event :my_event
+              with nil.
+
+              Events received:
+              {:my_other_event=>[{}]}
+            MESSAGE
+          end
+      end
+
+      it 'does not raise if event was logged 1x' do
+        track_event.call
+        expect { expect(subject).to(expectation.call) }.
+          not_to raise_error(RSpec::Expectations::ExpectationNotMetError)
+      end
+
+      it 'does not raise if event was logged 2x' do
+        track_event.call
+        track_event.call
+
+        expect { expect(subject).to(expectation.call) }.
+          not_to raise_error(RSpec::Expectations::ExpectationNotMetError)
+      end
+    end
+
+    context 'event name + hash' do
+      let(:track_event) { -> { subject.track_event :my_event, arg1: 42 } }
+      let(:track_event_with_different_args) { -> { subject.track_event :my_event, arg1: 43 } }
+      let(:track_other_event) { -> { subject.track_event :my_other_event } }
+      let(:expectation) { -> { have_logged_event :my_event, arg1: 42 } }
+
+      it 'raises if no event has been logged' do
+        expect { expect(subject).to(expectation.call) }.
```

Comment:
it took me a few times to understand what this was doing, and it's repeated a lot. do you think it would help readability if we pulled it into a let/helper?

```suggestion
        expect { expect_to_have_logged }.
```

Reply:
Yeah, it's awkward. What do you think about [this pattern](https://github.com/18F/identity-idp/pull/10328/files#diff-3ce0cb5616df30ded00a4ba048990c9017df28f8285c0e2aa5a8adb009c04a01R183-R191) (in ddcfe4d)?